asdf/commands/edit.py

```diff
@@ -0,0 +1,589 @@
+    if not is_yaml_file(oname):
+        print("A YAML file is expected, with '.yaml' extension.")
+        sys.exit(1)
+
+    # Write the YAML for the original ASDF file.
+    with open(oname, "wb") as ofd:
+        ofd.write(asdf_text)
+        ofd.write(yaml_text)
+
+    # Output message to user.
+    delim = "*" * 70
+    print(f"\n{delim}")
+    print("ASDF formatting and YAML schema validated.")
+    print(f"The text portion of '{fname}' is written to:")
+    print(f"    '{oname}'")
+    print(f"The file '{oname}' can be edited using your favorite text editor.")
+    print("The edited text can then be saved to the ASDF file of your choice")
+    print("using 'asdftool edit -s -f <edited text file> -o <ASDF file>.")
+    print(f"{delim}\n")
+
+    return
+
+
+def get_yaml_text_and_delimiter(yaml_text):
+    """
+    Splits the YAML text into the text and the end delimiter in preparation
+    for padding.
+    """
+
+    wdelim = b"\r\n...\r\n"
+    ldelim = b"\n...\n"
+    if yaml_text[-len(wdelim) :] == wdelim:
+        delim = wdelim
+    elif yaml_text[-len(ldelim) :] == ldelim:
+        delim = ldelim
+    else:
+        print("Unrecognized YAML delimiter ending the YAML text.")
+        print(f"It should be {wdelim} or {ldelim}, but the")
+        print(f"last {len(wdelim)} bytes are {yaml_text[-len(wdelim):]}.")
+        sys.exit(1)
+
+    return yaml_text[: -len(delim)], delim
+
+
+def pad_edited_text(edited_text, orig_text):
+    """
+    There is more text in the original ASDF file than in the edited text,
+    so we will pad the edited text with spaces.
+
+    Parameters
+    ----------
+    edited_text - The text from the edited YAML file
+    orig_text - The text from the original ASDF file
+
+    Return
+    ------
+    The padded text and the number of spaces added as pad.
+    """
+    diff = len(orig_text) - len(edited_text)
+
+    edited_text, delim = get_yaml_text_and_delimiter(edited_text)
+
+    padded_text = edited_text + b"\n" + b" " * (diff - 1) + delim
+    return padded_text, diff - 1
+
+
+def add_pad_to_new_text(edited_text, pad_size):
+    """
+    Adds pad to edited text.
+
+    Parameters
+    ----------
+    edited_text - The text from the edited YAML file.
+    pad_size - The number of spaces to add as a pad.
+
+    Return
+    ------
+    Pad text with the number of spaces requested as pad.
+    """
+
+    edited_text, delim = get_yaml_text_and_delimiter(edited_text)
+
+    pad = b" " * pad_size
+    padded_text = edited_text + b"\n" + pad + delim
+
+    return padded_text
+
+
+def write_block_index(fd, index):
+    """
+    Write the block index to an ASDF file.
+
+    Parameters
+    ----------
+    fd - The output file to write the block index.
+    index - A list of locations for each block.
+    """
+    if len(index) < 1:
+        return
+
+    # TODO - this needs to be changed to use constants.py and pyyaml
+    bindex_hdr = b"#ASDF BLOCK INDEX\n%YAML 1.1\n---\n"
```

Comment:
I mentioned this in my previous two reviews, but we should use pyyaml here and `constants.INDEX_HEADER`.